src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -759,9 +809,290 @@ subroutine dyn_mpas_init_phase3(self, cam_pcnst, pio_file)
+
+        if (.not. associated(num_scalars)) then
+            call self % model_error('Failed to find variable "num_scalars"', subname, __LINE__)
+        end if
+
+        ! While we are at it, check if its value is consistent.
+        if (num_scalars /= self % number_of_constituents) then
+            call self % model_error('Failed to allocate constituents', subname, __LINE__)
+        end if
+
+        call mpas_pool_add_dimension(self % domain_ptr % blocklist % dimensions, 'num_scalars', num_scalars)
+
+        nullify(mpas_pool)
+        nullify(num_scalars)
+
         call self % debug_print(subname // ' completed')
     end subroutine dyn_mpas_init_phase3
 
+    !-------------------------------------------------------------------------------
+    ! subroutine dyn_mpas_define_scalar
+    !
+    !> \brief  Defines the names of constituents at run-time
+    !> \author Michael Duda
+    !> \date   21 May 2020
+    !> \details
+    !>  Given arrays of constituent names and their corresponding waterness, which
+    !>  must have sizes equal to the number of constituents used to call
+    !>  `dyn_mpas_init_phase3`, this subroutine defines the scalars inside MPAS.
+    !>  Note that MPAS uses the term "scalar", but CAM-SIMA calls it "constituent".
+    !>  Furthermore, because MPAS expects all water scalars to appear in a
+    !>  contiguous index range, this subroutine may reorder the scalars to satisfy
+    !>  this constrain. Index mapping between MPAS scalars and constituent names
+    !>  can be looked up through `index_constituent_to_mpas_scalar` and
+    !>  `index_mpas_scalar_to_constituent`.
+    !> \addenda
+    !>  Ported and refactored for CAM-SIMA. (KCW, 2024-05-19)
+    !
+    !-------------------------------------------------------------------------------
+    subroutine dyn_mpas_define_scalar(self, constituent_name, is_water_species)
+        class(mpas_dynamical_core_type), intent(inout) :: self
+        character(*), intent(in) :: constituent_name(:)
+        logical, intent(in) :: is_water_species(:)
+
+        ! Possible CCPP standard names of `qv`, which denotes water vapor mixing ratio.
+        ! They are hard-coded here because MPAS needs to know where `qv` is.
+        ! Index 1 is exactly what MPAS wants. Others also work, but need to be converted.
+        character(*), parameter :: mpas_scalar_qv_standard_name(*) = [ character(strkind) :: &
+            'water_vapor_mixing_ratio_wrt_dry_air', &
+            'water_vapor_mixing_ratio_wrt_moist_air', &
+            'water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water' &
+        ]
+
+        character(*), parameter :: subname = 'dyn_mpas_subdriver::dyn_mpas_define_scalar'
+        integer :: i, j, ierr
+        integer :: index_qv, index_water_start, index_water_end
+        integer :: time_level
+        type(field3dreal), pointer :: field_3d_real => null()
+        type(mpas_pool_type), pointer :: mpas_pool => null()
```

Comment:
Same question here as earlier: should be just `nullify` these within the subroutine rather than implicitly saving them with initialization?

Reply:
Same as the above. I pushed https://github.com/ESCOMP/CAM-SIMA/pull/267/commits/a731e9769498b5a3b374c2329aaf97eaec2ff853 to resolve this.